File: models/bucket.go
package models

import "github.com/google/uuid"

type Bucket struct {
	UUID     uuid.UUID `json:"uuid" gorm:"type:uuid;primaryKey"`
	Name     string    `json:"name" gorm:"not null"`
	Type     string    `json:"type" gorm:"not null"`
	Size     int64     `json:"size" gorm:"not null;default:0"`
	RootNode *Node     `json:"root_node" gorm:"-:all"`
	MaxSize  *int64    `json:"max_size"`
	Users    []User    `json:"users" gorm:"many2many:user_buckets;"`
}

type UserBucket struct {
	BucketUUID uuid.UUID `json:"bucket_uuid" gorm:"not null;type:uuid"`
	Bucket     Bucket    `json:"bucket" gorm:"foreignKey:BucketUUID;not null"`
	UserID     int       `json:"user_id" gorm:"not null"`
	User       User      `json:"user" gorm:"foreignKey:UserID;not null"`
	AccessType string    `json:"access_type" gorm:"not null"`
}

type AccessType int

const (
	Denied = iota
	ReadOnly
	Write
	Full
)

File: src/main/kotlin/com/quentinguidee/services/Bucket.kt
package com.quentinguidee.services

import com.quentinguidee.models.db.*
import org.jetbrains.exposed.sql.and
import org.jetbrains.exposed.sql.insert
import org.jetbrains.exposed.sql.select
import org.jetbrains.exposed.sql.transactions.transaction
import java.util.*

class BucketService {
    suspend fun bucket(userID: Int) = transaction {
        val query = Buckets
            .innerJoin(UserBuckets)
            .innerJoin(Users)
            .select {
                Users.id eq userID and
                        (Buckets.type eq BucketType.USER_BUCKET)
            }
            .first()

        val bucket = Bucket.wrapRow(query)

        bucket.rootNode = Node
            .find { Nodes.bucket eq bucket.id and Nodes.parent.isNull() }
            .first()

        return@transaction bucket
    }

    suspend fun createBucket(userID: Int) =
        createBucket(userID, "User bucket", BucketType.USER_BUCKET)

    private suspend fun createBucket(userID: Int, name: String, type: BucketType) = transaction {
        val bucket = Bucket.new {
            this.name = name
            this.type = type
        }

        UserBuckets.insert {
            it[accessType] = AccessType.ADMIN
            it[user] = userID
            it[UserBuckets.bucket] = bucket.id
        }

        val rootNode = Node.new {
            this.bucket = bucket
            this.name = "root"
            this.type = "directory"
        }

        bucket.rootNode = rootNode

        return@transaction bucket
    }

    private suspend fun accessType(bucketUUID: UUID, userID: Int) = transaction {
        val query = UserBuckets
            .select {
                UserBuckets.user eq userID and
                        (UserBuckets.bucket eq bucketUUID)
            }
            .firstOrNull() ?: return@transaction null

        return@transaction query.let {
            it[UserBuckets.accessType]
        }
    }

    suspend fun authorize(desiredAccessType: AccessType, bucketUUID: UUID, userID: Int): Boolean {
        val accessType = accessType(bucketUUID, userID) ?: return false
        println("${accessType.ordinal} ${desiredAccessType.ordinal}")
        return accessType >= desiredAccessType
    }
}

val bucketService = BucketService()
